### ISSUE TITLE
[ops] ttnn.polygamma: silently wrong for x < 0; Euler-Maclaurin tail evaluated outside its domain

### ISSUE THREAD

**Issue by ssw-dot** [tenstorrent/tt-metal#52278]
### Component / Area

ops, kernels

### Issue Type (optional)

None

### Observed

`ttnn.polygamma` applies its Euler-Maclaurin tail correction at a fixed offset `z = x + NUM_TERMS` with `NUM_TERMS = 6`. The asymptotic expansion behind that correction is only valid for large `z`. For negative inputs `z` collapses, and the op returns silently wrong finite values, including sign flips, with no validation and no documented range.

`polygamma` is mathematically well defined for every non-integer `x`, including negative ones.

## Summary

The kernel (`ckernel_sfpu_polygamma.h`, byte-identical on Wormhole and Blackhole) computes

    psi^(n)(x) = (-1)^(n+1) * n! * [ sum_{k=0}^{5} 1/(x+k)^(n+1) + tail(z) ],  z = x + 6

where `tail(z)` is an Euler-Maclaurin remainder with B2, B4, B6 corrections. I verified the Bernoulli coefficients term by term against the Hurwitz-zeta Euler-Maclaurin expansion - the math is correct, and on the domain the kernel comment claims to support, accuracy is excellent.

The problem is that `z = x + 6` is a fixed shift. The expansion needs `z` comfortably large. As `x` goes negative, `z` shrinks toward zero and the expansion leaves its domain:

| x | z = x + 6 | consequence |
|---|---|---|
| -4.5 | 1.5 | degradation begins |
| -5.5 | 0.5 | asymptotic series diverges |
| -6.0 | 0.0 | division by zero |
| < -6 | negative | expansion evaluated on a negative base; signs invert |

## Measured

Reference values from `mpmath.polygamma` at 50 decimal digits. The kernel algorithm was ported faithfully to Python and evaluated in float64, so these are algorithmic errors - not rounding, and they do not improve in fp32.

**n = 1 (trigamma)**

| x | kernel | true | rel. error |
|---|---|---|---|
| -4.50 | 9.67074 | 9.67026 | 4.89e-05 |
| -5.50 | 12.0828 | 9.70332 | **24.5 %** |
| -6.50 | **-2.5221** | **+9.72699** | **126 % - sign flipped** |
| -8.50 | -0.111004 | 9.75861 | 101 % |
| -12.50 | -0.0768852 | 9.79272 | 101 % |

**n = 2 (tetragamma)**

| x | kernel | true | rel. error |
|---|---|---|---|
| -3.50 | -0.0615793 | -0.0615568 | 3.66e-04 |
| -4.50 | -0.0421896 | -0.0396089 | 6.5 % |
| -5.50 | **-35.1988** | -0.0275879 | **1,275x** |
| -6.50 | **-35.1915** | -0.0203053 | **1,733x** |

**Behaviour across the singularity at x = -6**, where `z` crosses zero:

| x | z | kernel | true |
|---|---|---|---|
| -5.90 | 0.10 | 2.34990e+05 | 103.2 |
| -5.99 | 0.01 | **2.38062e+12** | 1.00031e+04 |
| -6.01 | -0.01 | **-2.38062e+12** | 1.00031e+04 |

At `x = -6.01` the op returns a value off by eight orders of magnitude and with the wrong sign. Nothing raises, nothing warns.

## The positive domain is fine

To be clear about scope - I swept `n = 1..12` over `x` in [0.5, 200]:

    worst relative error across n=1..12, x >= 0.5 : 1.160e-09

The recent reduction from 11 exact terms to 6 is sound for the domain it targets. **This report is only about negative x.**

## Why this shipped

Every polygamma test samples strictly positive inputs:

| test | range |
|---|---|
| `tests/ttnn/unit_tests/operations/eltwise/test_composite.py:235` | `torch.rand(...) * 9.0 + 1.0` -> [1, 10] |
| `tests/ttnn/unit_tests/operations/eltwise/test_math.py:350` | `low = 1` |
| `tests/ttnn/nightly/.../test_backward_polygamma.py:32` | `data_gen_with_range(..., 1, 10, ...)` |
| `tests/ttnn/docs_examples/test_eltwise_unary_examples.py:1147` | `torch.tensor([[1, 2], [3, 4]])` |

No test evaluates a negative argument, so the failure is invisible to CI.

## API asymmetry

`unary_composite_op.cpp:339` validates the order:

    TT_FATAL(k >= 1 && k <= 11, "polygamma order must be in range [1, 11], got {}", k);

There is no corresponding validation of `x`, and no documented range in the binding. The kernel's own comment asserts a domain that is nowhere enforced or surfaced to users:

    // For the supported domain (x >= 0.5) this puts z >= 6.5 ...

So one parameter is guarded and documented while the other is silently assumed - the same shape of contract gap as the `i0` finding (#50465).

## Suggested fixes

**Minimal** - make the assumption explicit. Extend the existing `TT_FATAL` to reject `x` outside the supported domain, and document the range in the binding, exactly as `i1` documents [-10, 10].

**Complete** - keep `z` inside the asymptotic region instead of assuming it:

- Clamp the tail evaluation point: compute the tail at `z' = max(z, 6.5)` and add the exact terms bridging `x` to `z'`.
- Follow `ckernel_sfpu_i1.h`'s two-region pattern: compute unconditionally, then `v_if` the out-of-domain lanes onto a reflected path.
- Return NaN at the poles (non-positive integers).

**Test** - extend the sweeps past zero. A range of [-10, 10] on any of the four tests above would have caught this.

## Reproduction

No Tenstorrent hardware is required - the failure is in the algorithm, not the silicon. The Python port of the kernel and the sweep that produced every number above can be attached on request.

### Expected

Either of the following would resolve the contract gap:

1. **Validate and document the domain.** If `x >= 0.5` is genuinely the supported range, enforce it the same way the order is enforced - extend the existing `TT_FATAL` in `unary_composite_op.cpp:339` - and state the range in the binding docstring, as `i1` does with its "[Validated range: -10 to 10]" note. An out-of-domain input should fail loudly rather than return a plausible-looking wrong number.

2. **Or support negative x correctly.** Keep the Euler-Maclaurin evaluation point inside its region of validity instead of assuming `x + 6` is large enough, and return NaN at the poles (non-positive integers).

What should not happen is the current behaviour: a public op returning finite, silently incorrect values - off by 1,275x at x = -5.5 for n = 2, and by eight orders of magnitude with an inverted sign at x = -6.01 - with no error, no warning, and no documented restriction telling the user not to go there.

### 1. Steps (exact commands)

No Tenstorrent hardware is required. The defect is in the algorithm, so it reproduces exactly by porting the kernel arithmetic to Python.

1. Read `tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_polygamma.h` (the Blackhole copy is byte-identical).

2. Port `calculate_polygamma()` verbatim to Python in float64: six exact terms `sum_{k=0}^{5} 1/(x+k)^(n+1)`, then the Euler-Maclaurin tail evaluated at `z = x + 6` with coefficients `1/n`, `(n+1)/12`, `-(n+1)(n+2)(n+3)/720`, `(n+1)(n+2)(n+3)(n+4)(n+5)/30240`, and the `0.5/z` term. Multiply by `(-1)^(n+1) * n!`.

3. Compare against `mpmath.polygamma(n, x)` with `mp.dps = 50`.

4. Evaluate at n = 1 and x in [-4.5, -5.5, -6.5, -8.5, -12.5], and at n = 1 and x in [-5.90, -5.99, -6.01] to see the behaviour as `z` crosses zero.

Observed relative errors: 4.89e-05 at x = -4.5, 24.5 % at x = -5.5, 126 % with a sign flip at x = -6.5, and 2.38e+12 vs a true 1.00e+04 at x = -5.99.

On device the same call is `ttnn.polygamma(input_tensor, k=1)` with any tile containing values below about -4. No test in the repo currently does this - every polygamma test samples x in [1, 10] - which is why CI does not surface it.

I can attach the full Python reproduction script if that is useful.

### 2. Input data / link or description

No dataset, config or external file is needed. The inputs are plain scalars.

**Source files under analysis:**

- `tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_polygamma.h`
- `tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_polygamma.h` (byte-identical)
- `ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp` (line 339, the `TT_FATAL` on `k`)

**Input values that trigger the defect** - any `x` below roughly -4, with order `k` in [1, 11]:

    n = 1:  x = -4.5, -5.5, -6.5, -8.5, -12.5
    n = 2:  x = -3.5, -4.5, -5.5, -6.5
    across the singularity:  x = -5.90, -5.99, -6.01   (n = 1)

**Reference used for ground truth:** `mpmath.polygamma(n, x)` with `mp.dps = 50`. `mpmath` installs from PyPI with no other dependency.

On device the equivalent call is `ttnn.polygamma(input_tensor, k=1)` on any tile containing values in those ranges.

I have a self-contained Python script (~120 lines, numpy + mpmath only) that ports the kernel and prints the full comparison table. Happy to attach it to this issue or open it as a gist if that is preferred.

### 3. Frequency

Always. The defect is deterministic, not intermittent - it is an algorithmic error in the kernel arithmetic, so it reproduces on every evaluation with x below roughly -4.

### 1. Software Versions

Analysis performed against the tt-metal source on `main` - specifically `tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_polygamma.h` and `ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp` as of the date of this report.

To be transparent about method: I did not run this on a device build. The kernel's arithmetic was ported verbatim to Python (float64) and compared against `mpmath.polygamma` at 50 decimal digits. Because the error is algorithmic rather than a precision or hardware artefact, it reproduces identically at any precision.

Host used for the analysis: Ubuntu 22.04, Python 3.10, mpmath 1.3.0.

### 2. Hardware Details

None - no Tenstorrent hardware was used for this analysis.

The affected kernel file is byte-identical between Wormhole and Blackhole, so the defect applies to both. Silicon confirmation before any fix is merged would of course be welcome, and I would defer to a maintainer with device access on that step.

### Is this a regression?

Unknown

### Priority

None
